src/app/qgisapp.cpp

```diff
@@ -4271,7 +4271,27 @@ void QgisApp::updateRecentProjectPaths()
                               : QFileInfo( recentProject.path ).completeBaseName(), QDir::toNativeSeparators( recentProject.path )
                             ).replace( "&", "&&" )
                       );
-    //action->setEnabled( QFile::exists( ( recentProject.path ) ) );
+
+    QgsProjectStorage *storage = QgsApplication::projectStorageRegistry()->projectStorageFromUri( recentProject.path );
+
+    if ( storage )
+    {
+      QString path = storage->filePath( recentProject.path );
```

Comment:
Should this also be resetting recentProject.path to match?

Reply:
Can you elaborate more @nyalldawson? I didn't understood your point.

Reply:
Actually sorry, I had the wrong place for this comment. It actual relates to QgsWelcomePage::showContextMenuForProjects, where the project path is used to open the file in explorer. We'd need similar handling there to use the GPKG file path...

Reply:
Thanks @nyalldawson 
Now the context menu is aware of the project type (last commit https://github.com/qgis/QGIS/pull/31386/commits/17b73bfe7628d950cdda9fc2f563ffb5da993dda)
If the geopackage file is missing, users can open the closed path (I got it almost for free from the existing code).